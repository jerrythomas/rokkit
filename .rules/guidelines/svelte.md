# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
