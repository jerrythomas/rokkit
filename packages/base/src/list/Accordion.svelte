<script>
	import { defaultFields } from '../constants'
	import ListItems from './ListItems.svelte'
	import Collapsible from '../items/Collapsible.svelte'

	let className = ''
	export { className as class }
	export let items = []
	export let fields = {}
	export let using = {}
	export let autoClose = false

	let activeItem = null
	let activeGroup = null

	$: fields = { ...defaultFields, ...fields }
	$: using = { collapsible: Collapsible, ...using }

	function handleToggle(event) {
		if (autoClose) {
			if (activeGroup && activeGroup !== event.detail && activeGroup.isOpen) {
				activeGroup.isOpen = false
			}
			activeGroup = event.detail
		}
	}
</script>

<accordion class="flex flex-col w-full select-none {className}">
	{#each items as item}
		<group class="flex flex-col">
			<svelte:component
				this={using.collapsible}
				bind:content={item}
				{fields}
				on:toggle={handleToggle}
			/>
			{#if item.isOpen}
				<ListItems
					bind:items={item[fields.data]}
					bind:activeItem
					{fields}
					{using}
					on:click
				/>
			{/if}
		</group>
	{/each}
</accordion>
