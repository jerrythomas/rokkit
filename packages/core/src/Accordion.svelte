<script>
	import { Text, Summary } from './items'
	import { defaultFields } from './constants'
	import ListItems from './ListItems.svelte'

	let className = ''
	export { className as class }
	export let items = []
	export let fields = {}
	export let using = {}
	export let autoClose = false
	export let value = null

	let activeGroup = null

	$: fields = { ...defaultFields, ...fields }
	$: using = { default: Text, ...using }

	function handleToggle(event) {
		if (autoClose) {
			if (
				activeGroup &&
				activeGroup !== event.detail &&
				activeGroup[fields.isOpen]
			) {
				activeGroup[fields.isOpen] = false
			}
			activeGroup = event.detail
		}
	}
</script>

<accordion class="flex flex-col w-full select-none {className}">
	{#each items as item}
		{@const hasItems = item[fields.data] && item[fields.data].length > 0}
		{@const itemFields = fields.fields ?? fields}

		<details class="flex flex-col" class:expanded={item[fields.isOpen]}>
			<Summary {fields} {using} bind:content={item} on:toggle={handleToggle} />
			{#if hasItems && item[fields.isOpen]}
				<ListItems
					bind:items={item[fields.data]}
					bind:value
					fields={itemFields}
					{using}
					on:select
				/>
			{/if}
		</details>
	{/each}
</accordion>
