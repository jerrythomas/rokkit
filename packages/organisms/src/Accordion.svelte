<script>
	import { createEmitter } from '@rokkit/core'
	// import { createEventDispatcher } from 'svelte'
	import { defaultFields } from '@rokkit/core'
	import { navigator } from '@rokkit/actions'
	import { Item, Summary } from '@rokkit/molecules'
	// import List from './List.svelte'
	import ListItems from './ListItems.svelte'

	// const dispatch = createEventDispatcher()

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {any} [fields]
	 * @property {any} [using]
	 * @property {boolean} [autoClose]
	 * @property {any} [value]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		items = $bindable([]),
		fields = $bindable({}),
		using = $bindable({}),
		autoClose = false,
		value = $bindable(null),
		...events
	} = $props()
	let cursor = $state([])

	let emitter = $derived(createEmitter(events, ['collapse', 'change', 'expand', 'click']))
	let internalFields = $derived({ ...defaultFields, ...fields })
	let internalUsing = $derived({ default: Item, ...using })

	function handle(event) {
		value = event.detail.node
		cursor = event.detail.path
		if (['collapse', 'expand'].includes(event.type)) {
			if (autoClose) {
				items.map((x) => {
					if (x !== value && x[internalFields.isOpen]) {
						x[internalFields.isOpen] = false
					}
				})
			}
			items = items
		}
		emitter[event.type](event.detail)
		// dispatch(event.type, event.detail)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<accordion
	class="flex w-full select-none flex-col {className}"
	tabindex="0"
	use:navigator={{
		items,
		fields: internalFields,
		enabled: true,
		indices: cursor
	}}
	onselect={handle}
	onmove={handle}
	onexpand={handle}
	oncollapse={handle}
>
	{#each items as item, index}
		{@const hasItems = item[internalFields.children] && item[internalFields.children].length > 0}
		{@const itemFields = { ...internalFields, ...(internalFields.fields ?? internalFields) }}

		<div
			id={'id-' + index}
			class="flex flex-col"
			class:is-expanded={item[internalFields.isOpen]}
			class:is-selected={item === value}
			data-path={index}
		>
			<Summary fields={internalFields} {using} bind:value={items[index]} />
			{#if hasItems && item[internalFields.isOpen]}
				<list class="flex w-full flex-shrink-0 select-none flex-col" role="listbox" tabindex="-1">
					<ListItems
						bind:items={item[internalFields.children]}
						bind:value
						fields={itemFields}
						{using}
						hierarchy={[index]}
						on:change
					/>
				</list>
				<!-- <List
					bind:items={item[fields.children]}
					bind:value
					fields={itemFields}
					{using}
					on:select
					hierarchy={[index]}
					tabindex="-1"
				/> -->
			{/if}
		</div>
	{/each}
</accordion>
