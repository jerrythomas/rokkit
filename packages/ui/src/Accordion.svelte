<script>
	import { equals } from 'ramda'
	import { createEmitter, noop, getKeyFromPath, getSnippet } from '@rokkit/core'
	import { NestedController } from '@rokkit/states'
	import { navigator } from '@rokkit/actions'
	import Summary from './Summary.svelte'
	import Item from './Item.svelte'
	import ListBody from './ListBody.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {import('@rokkit/core').FieldMapper} [mapping]
	 * @property {boolean} [autoCloseSiblings]
	 * @property {boolean} [multiselect]
	 * @property {any} [value]
	 */

	/** @type {Props} */
	let {
		class: classes = '',
		items = $bindable([]),
		value = $bindable(null),
		fields,
		autoCloseSiblings = false,
		multiselect = false,
		header = null,
		footer = null,
		empty = null,
		stub = null,
		extra,
		...events
	} = $props()

	let emitter = $derived(
		createEmitter(events, ['collapse', 'change', 'expand', 'click', 'select', 'move'])
	)
	let wrapper = new NestedController(items, value, fields, {
		events,
		multiselect,
		autoCloseSiblings
	})
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<rk-accordion
	class={classes}
	tabindex="0"
	use:navigator={{ wrapper }}
	onactivate={() => (value = wrapper.value)}
>
	{#if header}
		<rk-header>{@render header()}</rk-header>
	{/if}
	{#if items.length === 0}
		<rk-list-item role="presentation">
			{#if empty}
				{@render empty()}
			{:else}
				No items found.
			{/if}
		</rk-list-item>
	{/if}
	{#each items as item, index}
		{@const key = `${index}`}
		<div
			class="flex flex-col"
			class:is-expanded={wrapper.expandedKeys.has(key)}
			class:is-selected={wrapper.selectedKeys.has(key)}
			data-path={index}
		>
			<Summary
				bind:value={items[index]}
				{fields}
				expanded={wrapper.expandedKeys.has(key)}
				hasChildren={wrapper.hasChildren(key)}
			/>
			{#if wrapper.expandedKeys.has(key)}
				<rk-list role="listbox" tabindex="-1">
					<ListBody
						bind:items={items[fields.children]}
						bind:value
						fields={fields.fields ?? fields}
						{selected}
						{stub}
						onchange={emitter.change}
						{extra}
					/>
				</rk-list>
			{/if}
		</div>
	{/each}
	{#if footer}
		<rk-footer>{@render footer()}</rk-footer>
	{/if}
</rk-accordion>
