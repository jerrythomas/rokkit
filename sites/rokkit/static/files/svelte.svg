<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
