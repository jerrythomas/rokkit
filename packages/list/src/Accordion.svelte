<script>
	import { onMount } from 'svelte'
	import { defaultFields } from './constants'
	import List from './List.svelte'
	import Collapsible from './items/Collapsible.svelte'

	export let items = []
	export let fields = {}
	export let using = {}
	export let selected
	export let autoClose = false
	let previous = null
	let expanded = {}

	$: fields = { ...defaultFields, ...fields }
	$: using = { collapsible: Collapsible, ...using }

	onMount(() => {
		if (autoClose) {
			items = items.map((item) => ({ ...item, collapsed: true }))
		}

		// expanded = items.reduce((acc, d) => ({ ...acc, [d[fields.id]]: false }), {})
	})

	function handleToggle(event) {
		console.log(event)
		if (autoClose) {
			if (previous && previous != event.detail && !previous.collapsed) {
				previous.collapsed = true
			}
			previous = event.detail
		}
	}
</script>

<accordion class="flex flex-col flex-shrink-0 w-full select-none">
	{#each items as item}
		<svelte:component
			this={using.collapsible}
			bind:content={item}
			{fields}
			on:toggle={handleToggle}
		/>
		{#if item[fields.data] && !item.collapsed}
			<List
				bind:items={item[fields.data]}
				{fields}
				{using}
				bind:selected
				on:click
				on:change
			/>
		{/if}

		<!-- <Collapsible
			id={parent[fields.groupId]}
			bind:name={parent[fields.text]}
			bind:items={parent[fields.data]}
			bind:selected
			bind:expanded={expanded[parent[fields.groupId]]}
			key={fields.itemId}
			icon={parent[fields.icon]}
			{component}
			on:expand={collapseOthers}
			on:select
			on:change
		/> -->
	{/each}
</accordion>
