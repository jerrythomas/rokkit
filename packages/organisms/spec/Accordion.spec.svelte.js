import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'

import Accordion from '../src/Accordion.svelte'

describe('Accordion.svelte', () => {
	beforeEach(() => cleanup())

	it('should render using default field mapping', () => {
		const { container } = render(Accordion)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	// it('should render using field mappings', () => {})
	// it('should render items using custom component', () => {})
	// it('should expand and collapse', () => {})
	// it('should autoclose others', () => {})
	// it('should pass select and change events', () => {})
})
