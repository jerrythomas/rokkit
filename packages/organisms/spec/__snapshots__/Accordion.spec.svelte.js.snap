// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion.svelte > should render using default field mapping 1`] = `
<div>
  <accordion
    class="flex w-full select-none flex-col "
    tabindex="0"
  >
    
  </accordion>
  
</div>
`;
